### PR TITLE
Heinous Junction Restrictions overlay

### LINKED ISSUE #633: Fix heinous junction restriction persistent overlay positions
Split from https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/621#issuecomment-581037030

While editing a node, the icons are in good positions, you can work out what's going on:

editing

However, the persistent overlay positioning is horrific:

overlay

Another example:

overlay

^ The enter blocked junction icon looks like it's for the 2-way road at the bottom, but it is in fact for the 1-way road entering from the right!

I suspect, but haven't checked, that this is some sort of artefact from trying to accommodate multiple overlays at same time (junction restricions, speed limits, priority signs, parking...); there was probably lack of space hence the dubious placement. In other words, I suspect that fixing the junction restrictions persistent overlay will then have side-effect of making them collide with some other overlay so this will probably turn in to can of worms.

### PR DESCRIPTION
**Pull request by kvakvs**
Fixes #633

# Problem

Different display style of Junction Restrictions overlay was very confusing see #633
⚠ Please note that the code is using segment middle point as the direction, and it is not a bug that on curved segments the signs look a bit off to the side. If this is undesired, i can calculate more precise endpoint tangent which will look better.

# Solution

* Refactored overlay code for Junction Restrictions
* Now readonly icons show as greyed out, smaller
* Now active icons show as full opaque, orange tint on hover
* Now icons always form a grid along the vector towards the segment center, always from the node position. Before: Was sometimes shifted left or right.
* The logic for which sign to be displayed: Not modified

bild

bild

### CODE REVIEW

**Comment by krzychu124** [comment]
Probably out of scope of that PR, but why we are showing pedestrian crossing icon if it's not possible to change its state? Highway roads doesn't have pedestrian lanes and also cannot have pedestrian crossing, obviously.

I'll try to test it later today.

**Comment by aubergine10** [comment]
This is looking a lot better, I can now clearly see which road the 'enter blocked junction' icons are relating to.

In this case it also seems to be putting them on the correct side of the road (left hand traffic map) which is awesome!

image

However, there's some cases where it seems to put the icons on the other side of the road, like this:

image

I thought it might be something to do with 1-way roads, but on same type of road it sometimes puts them in correct side like so:

image

It's only very minor point; I'm already happy with the big improvement from this PR and we could look at further refinements at a later date.

**Comment by kianzarrin** [comment]
>  Please note that the code is using segment middle point as the direction, and it is not a bug that on curved segments the signs look a bit off to the side. If this is undesired, i can calculate more precise endpoint tangent which will look better.

CS has calculated the middle position between two corners like this
```c#
//psudo-code:
pointA, tanA = NetSegment.CalculateCorner(junctionID, left side )
pointB, tanB = NetSegment.CalculateCorner(junctionID, right side)
centerPos = (postA+pointB)/2

//the original code can be find in this overloaded method:
//private void RefreshJunctionData(ushort nodeID, int segmentIndex, ushort nodeSegment, Vector3 centerPos, ref uint instanceIndex, ref RenderManager.Instance data)
```
you can use dirA and dirB from code above to align the signs. if you want to be even more thorough than that you should travel across curved road using `Bezier3.Travel`. I have done this is my pedestrian bridge builder: 
https://github.com/kianzarrin/PedestrianBridge/blob/d815c3920906965f7539f5d552ee45c45389a1c9/PedestrianBridge/Shapes/RoadBridgeWrapper.cs#L26 
and
https://github.com/kianzarrin/PedestrianBridge/blob/d815c3920906965f7539f5d552ee45c45389a1c9/PedestrianBridge/Shapes/RoadSideWrapper.cs#L69

I think calculating the tangent is going to be more complicated. 

In any case this is just in case you were interested. I don't think it is so important to put the signs at super accurate place.

**Comment by kianzarrin** [comment]
> bild

I don't know if this fits the scope of this review but there is no point to show pedestrian crossing overlays on roads without pedestrian crossing.

So `IsPedestrianCrossingAllowedConfigurable` should check if the segment does not have pedestrian crossing and the segments to the left and right also do not have pedestrian crossings then pedestrian crossing is not configurable. In such a case we can make user's job easier and not show the pedestrian crossings overlay to the user at all. 

In future when we have Pedestrian tool (#40) then we can  show a disabled pedestrian overlay and show a guide message to the user if he tries to click it. but here ... its too much info.

this is particularly important on roundabouts where there are a lot of traffic rules to set.

**Comment by kvakvs** [comment]
I had good results in the past when calculating tangent as a vector between 0.9f point and 1.0f point (or 0f and 0.1f if inverted direction).

```
if (this.debug_
                    || (configurable
                        && (!this.ViewOnly || !allowed))) {
```
The actual rendering code will render if debug is enabled. Should try on release configuration.

File: TLM/TLM/UI/SubTools/PrioritySigns/MassEditOverlay.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                }
+            }
+        }
+
+        private static DateTime _timer = DateTime.MinValue;
+
+        /// <summary>
+        /// show mass edit over lay for the input duration.
+        /// overrides MassEditOVerlay.Show when it is set to a UTC time in future.
+        /// seconds is
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/PrioritySigns/MassEditOverlay.cs:23]: can you delete these 2 lines?

Comment (kvakvs) [reply TLM/TLM/UI/SubTools/PrioritySigns/MassEditOverlay.cs]: Done

**Comment by kianzarrin** [comment]
Several issues that is probably outside of the scope of this review and exists in your branch as well as in master:
- if I turn junction restrictions overlay on and off from options while TMPE panel is open, mass edit  overlay stops working until I close and open TMPE panel 
- (your branch and master): activating the junction restriction tool turns off persistent overlay.
- I wish there was a option I could disable zebra crossings from persistent overlay because with my hide crossings mod, it is no longer all that necessary. maybe if hide crossing mod exists, then take out zebra crossings from persistent overlay?

**Comment by kvakvs** [comment]
@kianzarrin yes you can copy this comment later to the future PR when i'll be working on the LaneConnectors refactoring, it makes sense to fix those problems then.

**Comment by aubergine10** [comment]
Regarding the hide crossings mod: Note that some roads (eg. UK Roads assets) don't have zebra crossings so the only way to see if there is a crossing is via the overlay icons.

**Comment by kianzarrin** [comment]
> @aubergine10  Regarding the hide crossings mod: Note that some roads (eg. UK Roads assets) don't have zebra crossings so the only way to see if there is a crossing is via the overlay icons.

i said its no longer THAT necessary. meaning that it is necessary but not necessary enough to show it in persistent overlay. at the very least it would be nice to have the option to turn it off. 
we are creating a separate pedestrian tool in future so that might be the solution.
Also for highways, we can put a pedestrian prop so that it would not look weird when people cross the street. 
```
if road does not have 2+ pedestrian lanes. then add pedestrian crossing prop.
```
in any case there are a lot of urban roads without pedestrian crossings.

**Comment by kianzarrin** [comment]
looking better:

# OLD:
Screenshot (882)
Screenshot (883)

# VS NEW:
Screenshot (886)
Screenshot (885)

**Comment by kianzarrin** [comment]
@kvakvs I noticed that the sprites are always in the same place regardless of junction size:
Screenshot (867)
Screenshot (866)
Screenshot (885)

you should move the sprites further back as the junction gets bigger. so i think using `NetSegment.CalculateCorner() ` is the only way (and pretty easy too).

**Comment by kianzarrin** [comment]
Why is it the your sprites are overlapping? is that intentional?
Screenshot (874)
Screenshot (872)

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+                // Draw all junction restriction signs
+                // Determine direction from node center towards each segment center and use that
+                // as axis Y, and then dot product gives "horizontal" axis X
+                //------------------------------------
+                Vector3 segmentCenterPos = Singleton<NetManager>
+                                           .instance
+                                           .m_segments
+                                           .m_buffer[segmentId]
+                                           .m_bounds
+                                           .center;
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:247]: as I said before this is better be :
```
//psudo-code:
pointA, tanA = segment.CalculateCorner(junctionID, left side )
pointB, tanB = segment.CalculateCorner(junctionID, right side)
segmentCenterPos = (postA+pointB)/2
yBasis=(tanA+tanB ).normalized
```

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+            public SignsLayout(Vector3 nodePos,
+                               Vector3 yBasis,
+                               int signsPerRow,
+                               bool viewOnly,
+                               float baseZoom) {
+                this.signsPerRow_ = signsPerRow;
+                this.viewOnly_ = viewOnly;
+                baseZoom_ = baseZoom;
+
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:60]: `xBasis` becomes `yBasis` and `yBasis` becomes vector perpendicular to `yBasis`. Don't you think it would be better to use more consistent naming.
also this code is faster and much easier to read than cross product.(just a suggestion)
```C#
this.yBasis = new Vector3(-xBasis.z, 0, xBasis.x);// rotate 90 degree
```

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+            /// <param name="numSignsPerRow">How many signs per row.</param>
+            /// <param name="signSize">Sign size from <see cref="GetSignSize"/>.</param>
+            /// <returns>Zero position where first sign will go.</returns>
+            private static Vector3 GetZeroPosition(Vector3 nodePos,
+                                                   Vector3 xu,
+                                                   Vector3 yu,
+                                                   int numSignsPerRow,
+                                                   float signSize,
+                                                   bool viewOnly) {
+                // Vector3 centerStart = nodePos + (yu * (viewOnly ? 5f : 14f));
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:101]: commented out code

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+                baseZoom_ = baseZoom;
+
+                this.xBasis = yBasis;
+                this.yBasis = Vector3.Cross(yBasis, new Vector3(0f, 1f, 0f)).normalized;
+
+                this.signSize_ = viewOnly ? 3.5f : 6f;
+
+                this.zero_ = GetZeroPosition(
+                    nodePos: nodePos,
+                    xu: this.yBasis,
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:68]: again x and y are swapping!

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs (hunk 3 above)
Comment (kianzarrin) [reply TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:60]: you don't need the z dimension so y not to use 2D vectors? use `VectorUtils.XZ()` 
be aware of auto-casting `Vector2 v = vector3D` which is equivalent to `VectorUtils.XY()` and not what we want.

File: TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs (hunk 2 above)
Comment (kianzarrin) [reply TLM/TLM/UI/SubTools/PrioritySigns/Overlay.cs:247]: using `CalucluateCorner` would replace some of the calculations performed in `SignsLayout`.

**Comment by kvakvs** [comment]
@kianzarrin the core for calculating corners is not used anywhere in TM:PE, the source is massive like 5 pages of dense vector calculations. I almost matched everything to your suggestion but this will be called for each sign each frame...

**Comment by kvakvs** [comment]
Suggest @kianzarrin either add commit to this, or merge and come with your change later. I am literally down to experimentation if i try that corner calculation. That takes significant effort from me while you seem to know what is needed exactly.

**Comment by kianzarrin** [comment]
> @kianzarrin the C:S code for calculating corners is not used anywhere in TM:PE, there's no example of its meaning. Also the source is massive like 5 pages of dense vector calculations, it can't be fast or simple. I almost matched everything to your suggestion but this will be called for each segment each frame...
Maybe i can find a reasonable way of making it work and be fast.

@kvakvs sorry for the late response. I lost the review! In NetNodeRenderInstance corner is calculated in every call to render instance. I guess all those pages of calculation is not as slow as you think because there are so many branches and only one branch is taken. (i guess branching makes code even slower). In any case I think we should cache this stuff.

> Suggest @kianzarrin either add commit to this, or merge and come with your change later. I am literally down to experimentation if i try that corner calculation. That takes significant effort from me while you seem to know what is needed exactly.

Calculating the corner looks scare but it's disappointingly easy once you get to it (disappointing in case you want a challenge). Sure I can write some code and commit that in. 

BTW its nice that I can (for some reason) see the conflict. its helpful to know that its nothing.

**Comment by kianzarrin** [comment]
I simplified the code (I hope you don't mind!), changed the private interface a bit and deleted a few methods. also changed some variable names for example `zero_ -> origin_`.  Also fixed the x and y inconsistency (I removed so much code that x and y happen only once :) ). 

I use `CalculateCorner -> position and direction next to the crosswalks' for more precise calculations. 
Screenshot (903)

I defined some constants to shrink signs harmoniously. now the view only mode signs no longer overlap
Screenshot (904)

PS: the pictures above are a bit old. I moved the overlays a bit further up so that they don't cover the crosswalks.

**Comment by kianzarrin** [comment]
Screenshot (905)
Screenshot (906)

**Comment by kianzarrin** [comment]
In the picture bellow my FPS dropped from 25(no overlay) to 20 (a lot of overlay).
Screenshot (907)

This is far slower than @kvakvs code that drops FPS from 25 to 24

**Comment by kianzarrin** [comment]
Screenshot (908)
My FPS drops from 25 to 20 when I view the overlay
@kvakvs code drops FPS from 25 to 24.
Can you maybe cache the overlays?

**Comment by kianzarrin** [comment]
Wow my code is so clever it can even realise when the segment end is tilted:
Screenshot (909)

**Comment by kvakvs** [comment]
@kianzarrin you did not push the changes to my branch, did you push it anywhere else?

**Comment by kvakvs** [comment]
Not doing any caching at the moment.
This was intended to be a bugfix here not a huge rework :) Thanks @kianzarrin for your changes

File: TLM/TLM/UI/SubTools/JunctionRestrictionsTool.cs [hunk 6]
```diff
@@ -87,6 +87,10 @@ public JunctionRestrictionsTool(TrafficManagerTool mainTool)
             ushort updatedNodeId = 0;
             bool handleHovered = false;
             bool cursorInPanel = IsCursorInPanel();
+            PrioritySigns.PrioritySignsOverlay overlay = new PrioritySignsOverlay(
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/JunctionRestrictionsTool.cs:90]: Why did you rename `Overlay.cs` to `PrioirtySignsOverlay.cs` instead of moving it to `UI.Helpers` directory?
you call it `PrioirtySignsOverlay` but its only called by Junction Restriction tool (for now anyway).

**Comment by kianzarrin** [comment]
I added calculate corner cache, renamed priority signs overlay to traffic signs overlay (because it is only used for junction restriction for now) and changed a couple of comments.

dirX_ is not perpendicular to dirY_  (and I added a comment for that). hopefully that is not confusing. in any case there are axis systems without perpendicular axis so I think that is fine.

**Comment by kianzarrin** [comment]
Calculate corner is being cached to soon returning null.

**Comment by kianzarrin** [comment]
I added patch to Calculate corners after `CalculateSegment()` when the data is ready. It seems it takes a some while until segments are calculated.

**Comment by kianzarrin** [comment]
Created  #872 to discuss further when and where to perform recalculations.
